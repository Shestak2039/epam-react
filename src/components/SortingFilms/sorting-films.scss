@import "../../styles/colors";

.sorting-bar {
  width: 100%;

  display: flex;
  justify-content: space-between;

  border-bottom: 2px solid $secondary-color;
}
