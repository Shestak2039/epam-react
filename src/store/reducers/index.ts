import { combineReducers } from 'redux';

import filmsReducer from './films-reducer';

export default combineReducers({
    filmsState: filmsReducer
});
