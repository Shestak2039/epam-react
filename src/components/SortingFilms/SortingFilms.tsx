import React, { useState } from 'react';

import './sorting-films.scss';

import SortingByGenre, { GenreModel } from './SortingByGenre/SortingByGenre';
import Filter, { SortOptionModel } from './SortingByDate/Filter';

export interface SortingFilmsProps {
    sortFilms: (filter: GenreModel | undefined, sort: SortOptionModel | undefined) => void;
}

let mockGenres: GenreModel[] = [
    {id: 1, name: 'ALL', isSelected: true},
    {id: 2, name: 'DOCUMENTARY', isSelected: false},
    {id: 3, name: 'DRAMA', isSelected: false},
    {id: 4, name: 'HORROR', isSelected: false},
    {id: 5, name: 'THRILLER', isSelected: false}
];

let mockOptions: SortOptionModel[] = [
    {id: 1, name: 'NAME', isSelected: true},
    {id: 2, name: 'YEAR', isSelected: false},
];

const SortingFilms: React.FunctionComponent<SortingFilmsProps> = ({sortFilms}) => {
    const [genres, setGenres] = useState(mockGenres);

    const changeSelectedGenre = (newSelectedId: number) => {
        const newGenres = genres.map(genre => ({...genre, isSelected: newSelectedId === genre.id}));

        mockGenres = newGenres;
        setGenres(newGenres);

        sortFilms(newGenres.find(genre => genre.isSelected), options.find(option => option.isSelected));
    };

    const [options, setOptions] = useState(mockOptions);

    const changeSelectedOption = (newSelectedId: number) => {
        const newOptions = options.map(option => ({...option, isSelected: newSelectedId === option.id}));

        mockOptions = newOptions;
        setOptions(newOptions);

        sortFilms(genres.find(genre => genre.isSelected), newOptions.find(option => option.isSelected));
    };

    return (
        <div className="sorting-bar">
            <SortingByGenre genres={genres} setGenres={changeSelectedGenre}/>
            <Filter options={options} setOptions={changeSelectedOption}/>
        </div>
    );
};

export default SortingFilms;
