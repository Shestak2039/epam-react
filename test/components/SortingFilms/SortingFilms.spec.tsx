import React from 'react';
import { render } from '@testing-library/react';
import SortingFilms from '../../../src/components/SortingFilms/SortingFilms';
import { Provider } from 'react-redux';
import store from '../../../src/store/store';
import userEvent from '@testing-library/user-event/dist';


describe('SortingFilms', () => {
    test('renders SortingFilms snapshot', () => {
        const { asFragment, getByTestId, getByRole } = render(<Provider store={store}><SortingFilms /></Provider>);
        
        userEvent.click(getByTestId(1));
        userEvent.selectOptions(getByRole('listbox'), ['1']);
        
        expect(asFragment()).toMatchSnapshot();
    });
});
