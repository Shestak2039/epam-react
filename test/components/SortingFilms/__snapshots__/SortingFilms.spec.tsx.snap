// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SortingFilms renders SortingFilms snapshot 1`] = `
<DocumentFragment>
  <div
    class="sorting-bar"
  >
    <ul
      class="genres"
    >
      <li
        class="genre selected"
        data-testid="1"
      >
        ALL
      </li>
      <li
        class="genre"
        data-testid="2"
      >
        DOCUMENTARY
      </li>
      <li
        class="genre"
        data-testid="3"
      >
        DRAMA
      </li>
      <li
        class="genre"
        data-testid="4"
      >
        HORROR
      </li>
      <li
        class="genre"
        data-testid="5"
      >
        THRILLER
      </li>
    </ul>
    <div
      class="filter"
    >
      <div
        class="filter-name"
      >
        SORT BY
      </div>
      <select
        class="filter-select"
        role="listbox"
      >
        <option
          class="filter-select-option"
          selected=""
          value="1"
        >
          RATING
        </option>
        <option
          class="filter-select-option"
          value="2"
        >
          YEAR
        </option>
      </select>
    </div>
  </div>
</DocumentFragment>
`;
